conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java

```diff
@@ -472,43 +442,74 @@ private void trackAdEnd() {
-                    Log.d(TAG, "[Player Event] OnSourceUnloaded");
-                    internalEndSession();
-                }
-            }, 100);
+        public void onEvent(SourceEvent.Unloaded event) {
+        // The default SDK error handling is that it triggers the onSourceUnloaded before the onError event.
+        // To track errors on Conviva we need to delay the onSourceUnloaded to ensure the onError event is
+        // called first.
+        // TODO: remove this once the event order is fixed on the Android SDK.
+        new Handler().postDelayed(new Runnable() {
+            @Override
+            public void run() {
+                Log.d(TAG, "[Player Event] SourceUnloaded");
+                internalEndSession();
+            }
+        }, 100);
         }
     };
 
-    private OnErrorListener onErrorListener = new OnErrorListener() {
+    private final EventListener<PlayerEvent.Error> onPlayerErrorListener = new EventListener<PlayerEvent.Error>() {
         @Override
-        public void onError(ErrorEvent errorEvent) {
-            Log.d(TAG, "[Player Event] OnError");
+        public void onEvent(PlayerEvent.Error event) {
+            Log.d(TAG, "[Player Event] Error");
             try {
-                ensureConvivaSessionIsCreatedAndInitialized();
+                ConvivaAnalytics.this.ensureConvivaSessionIsCreatedAndInitialized();
 
-                String message = String.format("%s - %s", errorEvent.getCode(), errorEvent.getMessage());
+                String message = String.format("%s - %s", event.getCode(), event.getMessage());
                 client.reportError(sessionId, message, Client.ErrorSeverity.FATAL);
-                internalEndSession();
+                ConvivaAnalytics.this.internalEndSession();
+            } catch (ConvivaException e) {
+                Log.e(TAG, e.getLocalizedMessage());
+            }
+        }
+    };
+
+    private final EventListener<SourceEvent.Error> onSourceErrorListener = new EventListener<SourceEvent.Error>() {
+        @Override
+        public void onEvent(SourceEvent.Error event) {
+            Log.d(TAG, "[Source Event] Error");
+            try {
+                ConvivaAnalytics.this.ensureConvivaSessionIsCreatedAndInitialized();
+
+                String message = String.format("%s - %s", event.getCode(), event.getMessage());
+                client.reportError(sessionId, message, Client.ErrorSeverity.FATAL);
+                ConvivaAnalytics.this.internalEndSession();
+            } catch (ConvivaException e) {
+                Log.e(TAG, e.getLocalizedMessage());
+            }
+        }
+    };
+
+    private final EventListener<PlayerEvent.Warning> onPlayerWarningListener = new EventListener<PlayerEvent.Warning>() {
+        @Override
+        public void onEvent(PlayerEvent.Warning warningEvent) {
+            Log.d(TAG, "[Player Event] Warning");
+            try {
+                ensureConvivaSessionIsCreatedAndInitialized();
```

Comment:
Shouldn't this be `ConvivaAnalytics.this. ensureConvivaSessionIsCreatedAndInitialized();` as well? If it's not necessary, we might want to remove it in the `onSourceErrorListener`. Seems inconsistent.

Reply:
Good point, any input for us here @jarhoax ?

Reply:
Although there should be no ambiguity in names between anonymous event handler inner classes and outer class, but  using `ConvivaAnalytics.this.` sounds like a cleaner approach to prevent any ambiguity in future as well.

Reply:
On second thoughts, to make code consistent, the ConvivaAnalytics.this. prefix will need to be added at many places in all event handler classes. I will rather remove it from a few places where it is used.

Reply:
updated